### ISSUE TITLE
Translate VCF file into Tabular Format

### ISSUE THREAD

**Issue by noahaus** [Legall-Lab/SHAP-mTB-AMR#4]
<html>
<body>
<html><head></head><body><h1 data-path-to-node="2" style="font-family: &quot;Google Sans&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">VCF to ML-Ready Feature Matrix Conversion</h1><h2 data-path-to-node="3" style="font-family: &quot;Google Sans&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">📝 Description</h2><p data-path-to-node="4" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">After the generation of the <b data-path-to-node="4" data-index-in-node="28" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">VCF (Variant Call Format)</b> file, there is a requirement to transform the data into a structured format suitable for machine learning. This involves pivoting the genomic data so that variants serve as features and samples serve as observations.</p><h2 data-path-to-node="5" style="font-family: &quot;Google Sans&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">🛠 Technical Requirements</h2><p data-path-to-node="6" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">The transformation must result in a tabular dataset with the following specifications:</p><ul data-path-to-node="7" style="padding-inline-start: 32px; font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><li style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="7,0,0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="7,0,0" data-index-in-node="0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Columns:</b> Represent the specific <b data-path-to-node="7,0,0" data-index-in-node="32" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">genomic variants</b> collected (e.g., SNPs, Indels).</p></li><li style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="7,1,0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="7,1,0" data-index-in-node="0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Rows:</b> Represent individual <b data-path-to-node="7,1,0" data-index-in-node="27" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Samples</b>.</p></li><li style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="7,2,0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><b data-path-to-node="7,2,0" data-index-in-node="0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Target Labels:</b> Specific columns must capture <b data-path-to-node="7,2,0" data-index-in-node="45" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">Drug Resistance/Susceptibility</b> for the specific drugs under investigation.</p></li></ul><h2 data-path-to-node="8" style="font-family: &quot;Google Sans&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">🎯 Definition of Done</h2><ul data-path-to-node="9" style="padding-inline-start: 32px; font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><li style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="9,0,0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">[ ] A reproducible script/pipeline exists to perform the conversion.</p></li><li style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="9,1,0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">[ ] The output dataset is compatible with standard ML libraries (e.g., Scikit-Learn, XGBoost, or PyTorch).</p></li><li style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="9,2,0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">[ ] Target labels (Resistance/Susceptibility) are correctly mapped to each sample.</p></li><li style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;"><p data-path-to-node="9,3,0" style="font-family: &quot;Google Sans Text&quot;, sans-serif !important; line-height: 1.15 !important; margin-top: 0px !important;">[ ] The final matrix is validated for missingness and consistent dimensions.</p></li></ul>

</body></html>
</body>
</html>

**Comment by noahaus** [comment]
@nanzhen102 I was giving this a look recently and got an indication that the file might be wayyyyy too large after one hot encoding the sequences. I think it might be more advantageous to see if maybe we can look at (i) gene coding snps and then (ii) variations occuring in known AMR regions

**Comment by nanzhen102** [comment]
@noahaus 
Thanks. Do you think the following steps are reasonable?

**all_variants.csv.gz** (705,226 positions, 9,842 samples)

- [x]     **remove outliers** (e.g. ERR1034652 with 228,929 variants)
- [x]     **keep coding SNPs only** (~200,000 positions)
- Use Python to parse the H37Rv reference GFF to get [gene_start, gene_end] for every gene.
- [x]     **keep AMR genes only** (~5,000–15,000 positions)
- Start with drugs ***RIFAMPICIN, ISONIAZID, ETHAMBUTOL, PYRAZINAMIDE*** 
- Genes rpoB, katG, inhA, fabG1, embB, embA, embC, pncA, rpsA.
- [x]     **prevalence ≥1%** of samples (~3,000–10,000 positions)
- [x]     **pivot**: long to wide matrix (use pandas pivot_table)
- rows = samples
- columns = positions (e.g. pos_761155)
- values: 0=REF, 1=A, 2=T, 3=C, 4=G
- missing (no variant called) = 0 (assume REF)
- [x]     **join labels** via biosample map 
- ISONIAZID, RIFAMPICIN, ETHAMBUTOL, PYRAZINAMIDE
- R=1, S=0, NaN=unknown
- [x]     **ML-ready matrix**
- Random Forest (Use scikit-learn)
- SHAP (Use TreeExplainer)  
- [x]     **Which positions drive resistance predictions.**

**Comment by noahaus** [comment]
Yeah that sounds great! For the one hot encoding - the nucleotides can be 0
for reference, 1 for A, 2 for T, etc.

Sad that we can’t really utilize the indels- arguably the more interesting
variants, but we can maybe look at it later.

Even then, this is AMAZING progress - even the generation of the dataset is
amazing and we should maybe discuss ways to make it accessible (I can see
you Aspen announcing that this RIPE dataset is available for anyone to play
with to get your name out there 😌)

**Comment by nanzhen102** [comment]
@noahaus I updated some results. Please let me know your thoughts. 

**Random Forest + SHAP (v2. `scripts/rf_shap_v2.py`. Add balanced classes, AUC-ROC, 5-fold CV)**
| Drug | Accuracy | AUC-ROC | CV AUC-ROC |
|---|---|---|---|
| RIFAMPICIN | 95.0% | 0.975 | 0.969 ± 0.004 |
| ISONIAZID | 90.5% | 0.948 | 0.946 ± 0.008 |
| ETHAMBUTOL | 81.2% | 0.894 | 0.900 ± 0.007 |
| PYRAZINAMIDE | 81.9% | 0.886 | 0.883 ± 0.007 |

SHAP summary plots and per-drug metrics are in the `results/` folder. Will also upload a short paragraph about what we have done.

**Comment by noahaus** [comment]
This is great Aspen!  
I think once you make the paragraph, we can look into closing this issue and merging the work back into main.
Do you think we can put the dataset you used in here? I can't easily find it. Once that's done let's merge the branch you were working with to main and we can also create a new release to mark a new point in the project 😃

**Comment by nanzhen102** [comment]
@noahaus Sure! The dataset,` /resistance_dataset/ml_matrix.csv.gz` was committed. Sorry that I forgot to include it yesterday.
